### DISCUSSION TITLE
Pynetbox - get clusters of a specific type and tenant

### DISCUSSION THREAD

**Discussion by nomad-cyanide** [netbox-community/netbox#8910] [Q&A] [answered]
Hi

I am trying to get a list of the clusters of a specific type with a specific tenant set.

Type I get like this:
nb_clu_type=nb.virtualization.cluster_types.get(name='Check Point VSX')

Then I want to get the list of clusters, but nothin I have tried so far has worked, like
nb_clu=nb.virtualization.cluster.get(cluster_type_id=nb_clu_type.id)
nb_clu=nb.virtualization.cluster.get(type=nb_clu_type.id)
nb_clu=nb.virtualization.cluster.get(cluster_type=nb_clu_type.name)

The end goal is to get all the virtual machines in a list, from the clusters of type "Check Point VSX" assigned to a specific tenant.

What am I not understanding correctly?

Best regards
Ian

**Comment by candlerb** [comment]
Since you are trying to get a list of things, rather than a single thing, I think you want [`filter(...)`](https://pynetbox.readthedocs.io/en/latest/endpoint.html#pynetbox.core.endpoint.Endpoint.filter) rather than [`get(...)`](https://pynetbox.readthedocs.io/en/latest/endpoint.html#pynetbox.core.endpoint.Endpoint.get)

```
>>> list(nb.virtualization.virtual_machines.filter(cluster_type='ganeti'))
[elk]
>>>
```

**Reply by nomad-cyanide** [reply]
But this will not limit the search to only the specific cluster type within a specific tenant.

**Reply by candlerb** [reply]
It's up to you to modify the filter as appropriate; you can include multiple conditions in the query.  Note that you can either assign tenant to the *cluster* or to individual *virtual machines* so it depends exactly what you're trying to filter on.

At worst, you may have to do one query to find a list of clusters of interest, and then provide the list of clusters to the VM search:

```
list(nb.virtualization.virtual_machines.filter(cluster_id=[5,6]))
```

**Reply by nomad-cyanide** [reply]
I don't think I have been clear enough about my problem. First, I get that I should use filter, to get multiple items. That was a brainfart.
But really, I need to get the a list of the cluster of type X within tenant Z, not just any cluster with type X. This is the part that is troubling me.
I thought something like this might work: 
`nb_clu_tenant = list(nb.virtualization.cluster.filter(cluster_type_id=nb_clu_type.id,tenant_id=666)) `
but no such luck.

After I get that list of clusters, I need to list all of the VM's in those clusters, which likely would be something like:
```
for list in nb_clu_tenant:
   vms = nb.virtualization.virtual_machines.filter(cluster_id=cluster.id)
```

**Reply by candlerb** [reply]
Going back to the original query:

> I am trying to get a list of the clusters of a specific type with a specific tenant set

```
>>> nb.tenancy.tenants.get(name="blah").id
4
>>> list(nb.virtualization.clusters.filter(tenant_id=4))
[XXX, YYY, ZZZ]
>>> nb.virtualization.cluster_types.get(name="Proxmox VE").id
8
>>> list(nb.virtualization.clusters.filter(tenant_id=4,type_id=8))
[XXX, ZZZ]
```

> The end goal is to get all the virtual machines in a list, from the clusters of type "Check Point VSX" assigned to a specific tenant.

```
>>> cluster_ids = [c.id for c in nb.virtualization.clusters.filter(tenant_id=4,type_id=8)]
>>> cluster_ids
[68, 70]
>>> for v in nb.virtualization.virtual_machines.filter(cluster_id=cluster_ids): print(v)
...
VM1
VM2
...
```

NOTE: you can currently filter directly using **slugs**: here with Netbox v3.1.9 and pynetbox 6.6.1

```
>>> list(nb.virtualization.clusters.filter(type="proxmox-ve"))
[XXX, YYY, ZZZ, ...]
```

However my current understanding is that (sadly) slugs are going to be completely removed from Netbox in the near future, so you would be best advised not to do this.

EDIT: correction, they're not going away completely, but they may be weakened to the point where they are no longer usable as unique identifiers.  See #8113.

**Reply by candlerb** [reply]
> I thought something like this might work:
> `nb_clu_tenant = list(nb.virtualization.cluster.filter(cluster_type_id=nb_clu_type.id,tenant_id=666))`
> but no such luck.

Since the query is on cluster, you need `type_id` not `cluster_type_id`

**Comment by markkuleinio** [comment]
My VMs:

```
>>> for vm in nb.virtualization.virtual_machines.filter("VM"):
...     print(vm.name, vm.cluster)
...
TestVM Host1
VM1 Cluster1
VM2 Cluster2
```

My clusters and cluster types:
```
>>> for cluster in nb.virtualization.clusters.filter("Cluster"):
...     print(cluster.name, cluster.type)
...
Cluster1 Type1
Cluster2 Type2
```

VMs in clusters of Type1:
```
>>> nb_clu_type = nb.virtualization.cluster_types.get(name="Type1")
>>> clusters = nb.virtualization.clusters.filter(type_id=nb_clu_type.id)
>>> for cluster in clusters:
...     print("Cluster:", cluster)
...     print("VMs:", list(nb.virtualization.virtual_machines.filter(cluster_id=cluster.id)))
...
Cluster: Cluster1
VMs: [VM1]
```

**Comment by markkuleinio** [comment, marked as answer]
With tenant:

```
>>> for cluster in nb.virtualization.clusters.filter("Cluster"):
...     print(cluster.name, cluster.type, cluster.tenant)
...
Cluster1 Type1 None
Cluster1-Tenant1 Type1 Tenant1
Cluster2 Type2 None
>>> nb_clu_type = nb.virtualization.cluster_types.get(name="Type1")
>>> tenant = nb.tenancy.tenants.get(name="Tenant1")
>>> clusters = nb.virtualization.clusters.filter(type_id=nb_clu_type.id, tenant_id=tenant.id)
>>> for cluster in clusters:
...     print("Cluster:", cluster)
...     print("VMs:", list(nb.virtualization.virtual_machines.filter(cluster_id=cluster.id)))
...
Cluster: Cluster1-Tenant1
VMs: [VM3]
```

**Reply by nomad-cyanide** [reply]
`clusters = nb.virtualization.clusters.filter(type_id=nb_clu_type.id, tenant_id=tenant.id)`
Thank you. I think I have tried this already, but I'll try again. I may have made a mess in something else, since it didn't work for me yesterday. I'll get back to you.
